[JSC] Optimize runtime iterations for arguments and string

Constellation opened the pull request:
#### b5cc1fa8fff5886b7bf430b824101214da06fe9d
<pre>
[JSC] Optimize runtime iterations for arguments and string
<a href="https://bugs.webkit.org/show_bug.cgi?id=246714">https://bugs.webkit.org/show_bug.cgi?id=246714</a>
rdar://101317088

Reviewed by Alexey Shvayka.

This patch enhances the existing C++ runtime functions to efficiently iterate arguments and strings.
This change improves rollup with babel source code roughtly 8 - 10%.

Before:
    DYLD_FRAMEWORK_PATH=$VM $VM/jsc -m rollup.js  3.00s user 0.13s system 161% cpu 1.936 total
After:
    DYLD_FRAMEWORK_PATH=$VM $VM/jsc -m rollup.js  2.76s user 0.14s system 162% cpu 1.789 total

And some microbenchmarks get faster.

    array-slice-call-cloned-arguments           37.4436+-0.0669     ^     25.3087+-0.1296        ^ definitely 1.4795x faster
    cloned-arguments-varargs                     0.9464+-0.0349     ^      0.8217+-0.0278        ^ definitely 1.1518x faster
    spread-string                                5.5482+-0.0348     ^      3.1329+-0.0258        ^ definitely 1.7709x faster

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::sizeOfVarargs):
(JSC::loadVarargs):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::copyToArguments):
(JSC::ClonedArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/DirectArguments.cpp:
(JSC::DirectArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/DirectArguments.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastSlice):
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::toLength):
* Source/JavaScriptCore/runtime/JSCJSValue.cpp:
(JSC::JSValue::toLength const):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::objectPrototypeChainIsSane):
(JSC::JSGlobalObject::isArgumentsPrototypeIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromClonedArguments):
(JSC::createFromNonClonedArguments):
(JSC::JSImmutableButterfly::createFromDirectArguments):
(JSC::JSImmutableButterfly::createFromScopedArguments):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::maxSafeIntegerAsUInt64):
* Source/JavaScriptCore/runtime/ScopedArguments.cpp:
(JSC::ScopedArguments::unmapArgument):
(JSC::ScopedArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/ScopedArguments.h:

Canonical link: <a href="https://commits.webkit.org/255790@main">https://commits.webkit.org/255790@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/287f2e353a38529941276b302888a586e7566437

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2791 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103250 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163570 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2804 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31075 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85949 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99260 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80041 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84879 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37458 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79981 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35297 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/27615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39173 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82617 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1878 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41110 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/38000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/18679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

shvaikalesh commented on Source/JavaScriptCore/runtime/ScopedArguments.h:
```diff
@@ -161,6 +163,7 @@ class ScopedArguments final : public GenericArguments<ScopedArguments> {
     }
     
     bool m_overrodeThings { false }; // True if length, callee, and caller are fully materialized in the object.
+    bool m_unmapped { false };
```
nit: I would slightly prefer a bit more descriptive name like `m_hasUnmappedArgument`.

Constellation replied:
Sounds good, changed.

shvaikalesh reviewed (approved):
Marvellous, really like the cleanup around `toLength` return type.

Constellation committed:
[JSC] Optimize runtime iterations for arguments and string
https://bugs.webkit.org/show_bug.cgi?id=246714
rdar://101317088

Reviewed by Alexey Shvayka.

This patch enhances the existing C++ runtime functions to efficiently iterate arguments and strings.
This change improves rollup with babel source code roughtly 8 - 10%.

Before:
    DYLD_FRAMEWORK_PATH=$VM $VM/jsc -m rollup.js  3.00s user 0.13s system 161% cpu 1.936 total
After:
    DYLD_FRAMEWORK_PATH=$VM $VM/jsc -m rollup.js  2.76s user 0.14s system 162% cpu 1.789 total

And some microbenchmarks get faster.

    array-slice-call-cloned-arguments           37.4436+-0.0669     ^     25.3087+-0.1296        ^ definitely 1.4795x faster
    cloned-arguments-varargs                     0.9464+-0.0349     ^      0.8217+-0.0278        ^ definitely 1.1518x faster
    spread-string                                5.5482+-0.0348     ^      3.1329+-0.0258        ^ definitely 1.7709x faster

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::sizeOfVarargs):
(JSC::loadVarargs):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::copyToArguments):
(JSC::ClonedArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/ClonedArguments.h:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/DirectArguments.cpp:
(JSC::DirectArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/DirectArguments.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastSlice):
* Source/JavaScriptCore/runtime/JSArrayInlines.h:
(JSC::toLength):
* Source/JavaScriptCore/runtime/JSCJSValue.cpp:
(JSC::JSValue::toLength const):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::objectPrototypeChainIsSane):
(JSC::JSGlobalObject::isArgumentsPrototypeIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromClonedArguments):
(JSC::createFromNonClonedArguments):
(JSC::JSImmutableButterfly::createFromDirectArguments):
(JSC::JSImmutableButterfly::createFromScopedArguments):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.h:
* Source/JavaScriptCore/runtime/MathCommon.h:
(JSC::maxSafeIntegerAsUInt64):
* Source/JavaScriptCore/runtime/ScopedArguments.cpp:
(JSC::ScopedArguments::unmapArgument):
(JSC::ScopedArguments::isIteratorProtocolFastAndNonObservable):
* Source/JavaScriptCore/runtime/ScopedArguments.h:

Canonical link: https://commits.webkit.org/255790@main
